taroscript/send_test.go

```diff
@@ -877,59 +772,181 @@ func TestPrepareAssetCompleteSpend(t *testing.T) {
+	f: func(t *testing.T) error {
+		state := initSpendScenario(t)
+		state.spenderScriptKey = *asset.NUMSPubKey
 
-var prepareAssetCompleteSpendTestCases = []prepareAssetCompleteSpendTestCase{
-	{
-		name: "collectible with group key",
-		f: func(t *testing.T) error {
-			state := initSpendScenario(t)
-			spend := taroscript.SpendDelta{
-				InputAssets: state.
-					asset1CollectGroupInputAssets,
-			}
-			spendPrepared := taroscript.PrepareAssetCompleteSpend(
-				state.address1CollectGroup,
-				state.asset1CollectGroupPrevID, spend,
-			)
-			checkPreparedCompleteSpend(
-				t, spendPrepared, state.address1CollectGroup,
-				state.asset1CollectGroupPrevID,
-			)
-			return nil
-		},
-		err: nil,
+		pkt := createPacket(
+			state.address1, state.asset1PrevID,
+			state, state.asset1InputAssets,
+		)
+		err := taroscript.PrepareOutputAssets(pkt.Input, pkt.Outputs)
+		require.NoError(t, err)
+
+		pkt.Input.Asset().Genesis = state.genesis1collect
+		return taroscript.SignVirtualTransaction(
+			pkt.Input, pkt.Outputs, state.signer, state.validator,
+		)
 	},
-	{
-		name: "normal asset without split",
-		f: func(t *testing.T) error {
-			state := initSpendScenario(t)
-			spend := taroscript.SpendDelta{
-				InputAssets: state.asset1InputAssets,
-			}
-			spendPrepared := taroscript.PrepareAssetCompleteSpend(
-				state.address1, state.asset1PrevID, spend,
-			)
-			checkPreparedCompleteSpend(
-				t, spendPrepared, state.address1,
-				state.asset1PrevID,
-			)
-			return nil
-		},
-		err: nil,
+	err: vm.Error{Kind: vm.ErrIDMismatch},
+}, {
+	name: "validate with invalid NewAsset",
+	f: func(t *testing.T) error {
+		state := initSpendScenario(t)
+		state.spenderScriptKey = *asset.NUMSPubKey
+
+		pkt := createPacket(
+			state.address1, state.asset1PrevID,
+			state, state.asset1InputAssets,
+		)
+		err := taroscript.PrepareOutputAssets(pkt.Input, pkt.Outputs)
+		require.NoError(t, err)
+
+		firstPrevID := pkt.Outputs[0].Asset.PrevWitnesses[0].PrevID
+		firstPrevID.OutPoint.Index = 1337
+
+		return taroscript.SignVirtualTransaction(
+			pkt.Input, pkt.Outputs, state.signer, state.validator,
+		)
 	},
-}
+	err: vm.ErrNoInputs,
+}, {
+	name: "validate non-interactive collectible with group key",
+	f: func(t *testing.T) error {
+		state := initSpendScenario(t)
+		state.spenderScriptKey = *asset.NUMSPubKey
+
+		pkt := createPacket(
+			state.address1CollectGroup,
+			state.asset1CollectGroupPrevID, state,
+			state.asset1CollectGroupInputAssets,
+		)
+		err := taroscript.PrepareOutputAssets(pkt.Input, pkt.Outputs)
+		require.NoError(t, err)
 
-// TestCompleteAssetSpend tests edge cases around signing a witness for
-// an asset transfer and validating that transfer with the Taro VM.
-func TestCompleteAssetSpend(t *testing.T) {
+		unvalidatedAsset := pkt.Outputs[0].Asset.Copy()
+		err = taroscript.SignVirtualTransaction(
+			pkt.Input, pkt.Outputs, state.signer, state.validator,
+		)
+		require.NoError(t, err)
+
+		checkSignedAsset(
```

Comment:
Should be `true` since checking split root?

Reply:
Yes, correct, fixed.